src/dns-test.js

```diff
@@ -0,0 +1,298 @@
+                dnsData["tcp" + rrtype] = Array.from(responseBytes);
+            }
+            // If we didn't get an error, return.
+            // We don't need to re-transmit.
+            return;
+        } catch (e) {
+            let errorReason;
+            if (e.message.startsWith("STUDY_ERROR_TCP")) {
+                errorReason = e.message;
+            } else {
+                errorReason = STUDY_ERROR_TCP_MISC;
+            }
+            sendTelemetry({reason: errorReason,
+                           errorRRTYPE: rrtype,
+                           errorAttempt: dnsAttempts["tcp" + rrtype]});
+
+        }
+    }
+}
+
+/**
+ * Read the client's nameservers from disk.
+ * If on macOS, read /etc/resolv.comf.
+ * If on Windows, read a registry.
+ */
+async function readNameservers() {
+    let nameservers = [];
+    try { 
+        let platform = await browser.runtime.getPlatformInfo();
+        if (platform.os == "mac") {
+            nameservers = await browser.experiments.resolvconf.readNameserversMac();
+        } else if (platform.os == "win") {
+            nameservers = await browser.experiments.resolvconf.readNameserversWin();
+        } else {
+            throw new Error(STUDY_ERROR_NAMESERVERS_OS_NOT_SUPPORTED);
+        }
+    } catch(e) {
+        let errorReason;
+        if (e.message.startsWith("STUDY_ERROR_NAMESERVERS")) {
+            errorReason = e.message;
+        } else {
+            errorReason = STUDY_ERROR_NAMESERVERS_MISC;
+        }
+        sendTelemetry({reason: errorReason});
+        throw new Error(errorReason);
+    }
+
+    if (!(nameservers && nameservers.length)) {
+        sendTelemetry({reason: STUDY_ERROR_NAMESERVERS_NOT_FOUND});
+        throw new Error(STUDY_ERROR_NAMESERVERS_NOT_FOUND);
+    }
+    return nameservers;
+}
+
+/**
+ * For each RR type that we have a DNS record for, attempt to send queries over 
+ * UDP and TCP.
+ */
+async function sendQueries(nameservers_ipv4) {
+    for (let rrtype of RRTYPES) {
+        if (rrtype == 'SMIMEA') {
+            await sendUDPQuery(SMIMEA_DOMAIN_NAME, nameservers_ipv4, rrtype);
+            await sendTCPQuery(SMIMEA_DOMAIN_NAME, nameservers_ipv4, rrtype);
+        } else if (rrtype == 'HTTPS') {
+            await sendUDPQuery(HTTPS_DOMAIN_NAME, nameservers_ipv4, rrtype);
+            await sendTCPQuery(HTTPS_DOMAIN_NAME, nameservers_ipv4, rrtype);
+        } else {
+            await sendUDPQuery(APEX_DOMAIN_NAME, nameservers_ipv4, rrtype);
+            await sendTCPQuery(APEX_DOMAIN_NAME, nameservers_ipv4, rrtype);
+        }
+    }
+}
+
+/**
+ * Add an ID to telemetry that corresponds with this instance of our
+ * measurement, i.e. a browser session
+ */
+function sendTelemetry(payload) {
+    try {
+        payload.measurementID = measurementID;
+        browser.telemetry.submitPing(TELEMETRY_TYPE, payload, TELEMETRY_OPTIONS);
+    } catch(e) {
+        /* *
+         * Do nothing; we just don't want the addon to error out. The pings 
```

Comment:
The current construction here doesn't catch any telemetry errors.

`browser.telemetry.submitPing` is an asynchronous extension API, i.e. returns a `Promise`. If an error occurs, it will be printed to the console (because `.submitPing(..)` returns a promise, and an unhandled promise rejection may be logged to the console).
The only way for this snippet to throw is if the telemetry extension API is unavailable (e.g. when the add-on is mistakenly not signed as a privileged extension), or when the input doesn't match [the schema definition of the API](https://searchfox.org/mozilla-central/rev/23dd7c485a6525bb3a974abeaafaf34bfb43d76b/toolkit/components/extensions/schemas/telemetry.json#140-190). The latter is not possible, and the first should not happen (and wouldn't be reached because in that case the `browser.telemetry.canUpload` call at the start would have thrown), so the try-catch here is redundant.

I suggest to simply remove `try-catch` here.

Reply:
Addressed by 5d32dd7